modules/gui/src/com/haulmont/cuba/gui/data/impl/AbstractCollectionDatasource.java

```diff
@@ -644,28 +641,28 @@ protected void setSortDirection(LoadContext.Query q) {
 
     @Nullable
     protected String[] getSortPropertiesForPersistentAttribute(MetaPropertyPath propertyPath) {
-        String[] sortProperties = null;
+        ArrayList<String> props = new ArrayList<>();
         if (!propertyPath.getMetaProperty().getRange().isClass()) {
             // a scalar persistent attribute
-            sortProperties = new String[1];
-            sortProperties[0] = propertyPath.toString();
+            props.add(propertyPath.toString());
         } else {
             // a reference attribute
             MetaClass metaClass = propertyPath.getMetaProperty().getRange().asClass();
             if (!propertyPath.getMetaProperty().getRange().getCardinality().isMany()) {
-                InstanceUtils.NamePatternRec rec = InstanceUtils.parseNamePattern(metaClass);
-                if (rec != null) {
-                    sortProperties = new String[rec.fields.length];
-                    for (int i = 0; i < rec.fields.length; i++) {
-                        sortProperties[i] = propertyPath.toString() + "." + rec.fields[i];
-                    }
+                Collection<MetaProperty> properties = metadata.getTools().getNamePatternProperties(metaClass);

```

Comment:
Well, I dunno... We do not know the size of valid properties. So we must create another collection (ArrayList for instance), and with foreach loop add to this collection values from `Collection<MetaProperty> properties` and than create the result array. This is not more readable nor faster (as I correctly understand your concern).

Reply:
Since github is pointing me on previous commit, may be you don't seen the [last one](https://github.com/cuba-platform/cuba/pull/22/commits/3b8ff836a3cd085d4696fb2808463dc271425cf5)?

Reply:
OK, never mind.